content/4_Gold/PURS.mdx

```diff
@@ -160,241 +160,307 @@ public class segtree {
 </Resources>
 
 <LanguageSection>
-
 <CPPSection>
 
-Note that `st.init(n+1)` allows us to update and query indices in the range
-`[0,n+1)=[0,n]`.
-
 ```cpp
-#include <bits/stdc++.h>
-using namespace std;
+#include <iostream>
+#include <cassert>
+#include <vector>
+#include <algorithm>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+/** A data structure that can answer point update & range minimum queries. */
+template<class T>
+class MinSegmentTree {
+	private:
+		/** The operation to use for combining two elements. (Must be associative) */
+		T comb(T a, T b) { return std::min(a, b); }
+		const T DEFAULT = 1e18;  // Will overflow if T is an int
+
+		vector<T> segtree;
+		int len;
+	public:
+		MinSegmentTree(int len) : len(len), segtree(len * 2, DEFAULT) { }
+
+		/** Sets the value at ind to val. */
+		void set(int ind, T val) {
+			assert(0 <= ind && ind < len);
+			ind += len;
+			segtree[ind] = val;
+			for (; ind > 1; ind /= 2) {
+				segtree[ind >> 1] = comb(segtree[ind], segtree[ind ^ 1]);
+			}
+		}
 
-template<class T> struct Seg { // comb(ID,b) = b
-	const T ID = 1e18; T comb(T a, T b) { return min(a,b); }
-	int n; vector<T> seg;
-	void init(int _n) { n = _n; seg.assign(2*n,ID); }
-	void pull(int p) { seg[p] = comb(seg[2*p],seg[2*p+1]); }
-	void upd(int p, T val) { // set val at position p
-		seg[p += n] = val; for (p /= 2; p; p /= 2) pull(p); }
-	T query(int l, int r) {	// min on interval [l, r]
-		T ra = ID, rb = ID;
-		for (l += n, r += n+1; l < r; l /= 2, r /= 2) {
-			if (l&1) ra = comb(ra,seg[l++]);
-			if (r&1) rb = comb(seg[--r],rb);
+		/** @return the minimum element in the range [start, end) */
+		T range_min(int start, int end) {
+			assert(0 <= start && start < len && 0 < end && end <= len);
+			T sum = DEFAULT;
+			for (start += len, end += len; start < end; start /= 2, end /= 2) {
+				if ((start & 1) != 0) {
+					sum = comb(sum, segtree[start++]);
+				}
+				if ((end & 1) != 0) {
+					sum = comb(sum, segtree[--end]);
+				}
+			}
+			return sum;
 		}
-		return comb(ra,rb);
-	}
 };
 
-Seg<int> st;
-
 int main() {
-	int n, q; cin >> n >> q;
-	st.init(n+1);
-	for(int i=1; i<=n; i++) {
-		int a; cin >> a;
-		st.upd(i, a);
+	int arr_len;
+	int query_num;
+	std::cin >> arr_len >> query_num;
+
+	MinSegmentTree<int> segtree(arr_len);
+	for (int i = 0; i < arr_len; i++) {
+		int n;
+		std::cin >> n;
+		segtree.set(i, n);
 	}
-	for(int i=1; i<=q; i++) {
-		int t, a, b; cin >> t >> a >> b;
-		if (t==1) st.upd(a,b);
-		else cout << st.query(a,b) <<"\n";
+
+	for (int q = 0; q < query_num; q++) {
+		int type;
+		std::cin >> type;
+		if (type == 1) {
+			int ind, val;
```

Comment:
sep lines

Reply:
thought they were closely related enough, but alr then

Reply:
you separate them in the other langs, so might as well be consistent